File: aldStat_offline_cos/src/main/scala/aldwxutils/TimeUtil.scala
package aldwxutils

import java.sql.Timestamp
import java.text.{DateFormat, SimpleDateFormat}
import java.util.regex.Pattern
import java.util.{Calendar, Date, Locale}

import org.apache.commons.lang.StringUtils
import org.apache.commons.lang.time.DateFormatUtils


/**
  *  时间工具类
  */

object TimeUtil{
  //获得一天的时间(毫秒)
  val day= 24*3600*1000
  //获得一天时间（秒）
  val dayS= 24*3600

  var flag = false


  /**
  * <b>author:</b> gcs <br>
  * <b>data:</b> 18-5-7 <br>
  * <b>description:</b><br>
    *   自定义函数计算昨天日期 <br>
  * <b>param:</b><br>
  * <b>return:</b><br>
  */
  def ytime(): String = {
    //创建calendar对象
    val cal = Calendar.getInstance() //gcs:获得一个日期的对象
    cal.add(Calendar.DATE, -1) //gcs:让当前的日期减去1，就获得了昨天的日期

    val format = new SimpleDateFormat("yyyyMMdd")
    val ydate = cal.getTime  //gcs:获得日期
    val newydate = format.format(ydate) //gcs:使用格式化的日期对时间数据进行格式的操作
    return newydate //gcs:返回格式化之后的时间
  }

  // 返回今天的日期
  def ytoday(): String = {
    //创建calendar对象
    val cal = Calendar.getInstance()
    val format = new SimpleDateFormat("yyyyMMdd")
    val ydate = cal.getTime
    val newydate = format.format(ydate)
    return newydate
  }

  //自定义函数计算当天日期
  def time():String={
    //创建calendar对象
    val cal = Calendar.getInstance()
    DateFormat.getDateInstance.format(cal.getTime)
  }


  /**
  * <b>author:</b> gcs <br>
  * <b>data:</b> 18-5-4 <br>
  * <b>description:</b><br>
    *   将-d 中得到参数<br>
    *     传入的参数的格式是 "-d 2017-05-02" <br>
    *       为什么要有一个"-d 2017-05-02"的参数呢？比如我要补充2017-05-02的昨天，7天和30天的数据，我添加一个参数就可以将实现这个功能了。添加参数是提供了一个灵活的操作 <br>
  * <b>param:</b><br>
    *   args: Array[String] ; 在提交jar包时往里面传入的时间参数 <br>
  * <b>return:</b><br>
    *   String ;如果，args中有"-d 2017-05-03"，那么就会返回2017-05-03的数据20170503.
    *
  */
  def processArgs(args: Array[String]):String={
    var date = ""

    //==========================================================1
    /*gcs:
    *解析args传回来的天数 args中的参数格式是"-d 2017-05-03"
    */
    for (i <- 0 to args.length-1) {
      if ("-d".equals(args(i))) { //gcs:这里说明了,参数的类型是 -d 2017-05-03。这样找到-d 之后，-d后面的2017-05-03就可以被获取到了
        if (i + 1 < args.length) {
          date = args(i + 1)
        }
      }
    }



    if (StringUtils.isNotBlank(date) && chenkDate(date)){ //gcs:如果data不是null，并且符合规定的正则表达式
      val world = date.split("-")
      world(0)+world(1)+world(2)   //gcs:如果给的日期是 2017-05-03  这里的 world(0) 是2017 , world(1)是05,world(2)是03。这样的 world(0)+world(1)+world(2) 是20170503
    }else{
      ytime()  //gcs:如果参数中没有指定“-d 2-17-05-03”天数，就会自动获取昨天的时间
    }
  }

  /**
  * <b>author:</b> gcs <br>
  * <b>data:</b> 18-5-7 <br>
  * <b>description:</b><br>
    *   使用正则表达式对传进来的date信息进行判断。判断日期格式是否正确 <br>
  * <b>param:</b><br>
    *   date:String ;等待被判断的日期的格式 <br>
  * <b>return:</b><br>
    *   Boolen 判断传回来的日期是否符合正则表达式的格式 <br>
  */
  def chenkDate(date:String) :Boolean = {
    //定义正则表达式yyyyMMdd
    val regex = "[2][0][0-9]{2}-[0-9]{2}-[0-9]{2}"  //gcs:格式就像是 2017-05-03
    val compile =Pattern.compile(regex)
    //匹配数据
    val matcher =compile.matcher(date)
    matcher.matches()
  }


  //将 时间 转换成 long类型
  def long2int():Long={
    val sdf = new SimpleDateFormat("yyyy-MM-dd")
    val now = new Date()
    val dt =sdf.parse(sdf.format(now)).getTime-day
    val st =dt+""
    st.substring(0,10).toLong
  }

  //将long 类型转换成 string
  def long2Str(long: Long):String ={
    val sdf = new SimpleDateFormat("yyyyMMdd")
    val data = new Date(long)
    sdf.format(data)
  }


  def getDateString(millis: Long, pattern: String): String = {
    return DateFormatUtils.format(millis,pattern,Locale.ENGLISH)
  }


  def ytimeHive():String={
    //创建calendar对象
    val cal = Calendar.getInstance()
    cal.add(Calendar.DATE,-0)
    val format =new SimpleDateFormat("yyyyMMdd")
    val ydate =cal.getTime
    val newydate = format.format(ydate)
    newydate
  }

  def ytimeHive7():String={
    //创建calendar对象
    val cal = Calendar.getInstance()
    cal.add(Calendar.DATE,-7)
    getDateString(cal.getTime.getTime,"yyyyMMdd")
  }

  def ytimeHive29():String={
    //创建calendar对象
    val cal = Calendar.getInstance()
    cal.add(Calendar.DATE,-30)
    getDateString(cal.getTime.getTime,"yyyyMMdd")
  }

  /**
  * <b>author:</b> gcs <br>
  * <b>data:</b> 18-5-4 <br>
  * <b>description:</b><br>
    *   把String类型的时间类型数据转换成long类型的
  * <b>param:</b><br>
  * <b>return:</b><br>
  */
  def str2Long(str :String):Long= {
    val sdf = new SimpleDateFormat("yyyyMMdd")
    val date = sdf.parse(str)
    date.getTime
  }

  //当前时间的 int 类型
  def nowInt():Int={
    val sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    //创建calendar对象
    val cal = Calendar.getInstance()
    val now = new Date()
    val dt =sdf.parse(sdf.format(now)).getTime
    val st =dt+""
    st.substring(0,10).toInt
  }
  def StrToLong(str :String):Long= {
    val sdf = new SimpleDateFormat("yyyy-MM-dd")
    val date = sdf.parse(str)
    (date.getTime/1000).toLong
  }

  def str2int(string: String):Int={
    val sdf = new SimpleDateFormat("yyyyMMdd")
    //val now = new Date()
    val dt =sdf.parse(sdf.format(string)).getTime
    (dt/1000).toInt
  }


  /**
  * <b>author:</b> gcs <br>
  * <b>data:</b> 18-5-7 <br>
  * <b>description:</b><br>
    *   字符串转换Timestamp <br>
  * <b>param:</b><br>
    *   x:String ;string类型的时间 <br>
  * <b>return:</b><br>
    *   java.sql.Timestamp ;将String类型的时间转换成了Timestamp <br>
  */
  def getTimestamp(x:String) :java.sql.Timestamp = {

    val format = new SimpleDateFormat("yyyyMMdd") //gcs:设定一个时间格式
    var ts = new Timestamp(System.currentTimeMillis()) //gcs:获得当前时间的格式
    try {
      if (x == "")
        return null
      else {
        val d = format.parse(x); //gcs:转换时间
        val t = new Timestamp(d.getTime())
        return t
      }
    } catch {
      case e: Exception => println("cdr parse timestamp wrong")
    }
    return null
  }

  def todayStr():String={
    //创建calendar对象
    val cal = Calendar.getInstance()
    cal.add(Calendar.DATE,-1)
    getDateString(cal.getTime.getTime,"yyyyMMdd")
  }

  def processArgs2(args:Array[String]):StringBuffer= {
    var date = ""
    var date2 = ""
    for (i <- 0 to args.length - 1) {
      if ("-d".equals(args(i))) {
        if (i + 2 < args.length) {
          date = args(i + 1)
          date2 = args(i + 2)
        } else if (i + 1 < args.length) {
          date = args(i + 1)
        }
      }
    }
    var str:StringBuffer=new StringBuffer

    if (StringUtils.isNotBlank(date) && chenkDate(date)){
      flag = true
      val world = date.split("-")
      str.append(world(0)+ world(1) + world(2))

      if (StringUtils.isNotBlank(date2)){
        if (chenkDate(date2)){
          val world2 = date2.split("-")
          str.append(","+world2(0) + world2(1)+ world2(2))
        }else{
          str.append(","+ytime())
        }
      }else{
        str.append("")
      }
    }else{
      str.append(ytime())
    }
  }

  //获取 今天的long 类型 时间戳
  def dayLong():Long={
    val sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    //创建calendar对象
    val cal = Calendar.getInstance()
    val now = new Date()
    val dt =sdf.parse(sdf.format(now)).getTime
    val st =dt+""
    st.toLong
  }

  // 获取上周的周一的日期
  def getoldWeekStart():String={
    var period:String=""
    var cal:Calendar =Calendar.getInstance();
    var df:SimpleDateFormat = new SimpleDateFormat("yyyyMMdd");
    cal.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY)
    cal.add(Calendar.WEEK_OF_YEAR, -1)
    //获取上周一的日期
    period=df.format(cal.getTime())
    period

  }

  // 获取上周的周日的日期
  def getoldWeekEnd():String={
    var period:String=""
    var cal:Calendar =Calendar.getInstance();
    var df:SimpleDateFormat = new SimpleDateFormat("yyyyMMdd");
    cal.set(Calendar.DAY_OF_WEEK, Calendar.SUNDAY);//这种输出的是上个星期周日的日期，因为老外把周日当成第一天
    period=df.format(cal.getTime.getTime)
    period

  }

  //获得 当前的时间（小时数）
  def getHour():String={
    val now = new Date()
    val dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    val a = dateFormat.format(now.getTime()-3600000)
    val str =a.split(" ")
    str(1).substring(0,2)
  }

  def getdate():String={
    val now = new Date()
    val dateFormat = new SimpleDateFormat("yyyy-MM-dd")
    val a = dateFormat.format(now.getTime()-3600000)
    val str =a.split("-")
    str(0)+str(1)+str(2)
  }

  // 获取当前时间的时间戳
  def getTimeString(): String = {
    val df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
    val calendar = Calendar.getInstance
    val str = calendar.getTime.getTime+""
    str.substring(0,10)
  }


  //获得传入参数的时间戳
  def processArgs2Long(args: Array[String]) :Long ={
    val sdf = new SimpleDateFormat("yyyyMMdd")
    val date = processArgs(args)
    val day=date.substring(0,4)
    val hour=date.substring(4,6)
    val min = date.substring(6,8)

    val str = day+"-"+hour+"-"+min
    DateFormat.getDateInstance.parse(str).getTime
    //DateFormat.getDateInstance.parse(date).getTime
  }

  //判断传入的参数是不是日期格式(传参数)
  def processArgsStayTimeDay(args: Array[String]):String={
    var date = ""
    for (i <- 0 to args.length-1){
      if ("-d".equals(args(i))){
        if (i+1 < args.length){
          date =args(i+1)
        }
      }
    }
    if (StringUtils.isNotBlank(date) && chenkDate(date)){
      flag=true
      val world = date.split("-")
      world(0)+world(1)+world(2)
    }else{
      getdate()
    }
  }

  //判断 日期格式
  def chenkhour(date:String) :Boolean = {
    //定义正则表达式yyyyMMdd
    val regex = "[0-9]{2}"
    val compile =Pattern.compile(regex)
    //匹配数据
    val matcher =compile.matcher(date)
    matcher.matches()
  }

  //判断传入的参数是不是日期格式(传参数)  今日日期
  def processArgsStayTimeHour(args: Array[String]):String={
    var hour = ""
    for (i <- 0 to args.length-1){
      if ("-h".equals(args(i))){
        if (i+1 < args.length){
          hour =args(i+1)
        }
      }
    }
    if (StringUtils.isNotBlank(hour) && chenkhour(hour)){
      flag=true
      hour
    }else{
      getHour()
    }
  }


  def main(args: Array[String]): Unit = {
    println(long2int())
    println(processArgs2Long(args))
  }
}

File: aldwxStat_etl/aldwxstat_etl/etl/src/main/scala/aldwxutil/TimeUtil.java
package aldwxutil;

import org.apache.commons.lang.StringUtils;

import java.io.Serializable;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by zhangyanpeng on 2017/7/27.
 */
public class TimeUtil implements Serializable {

    //将 st 时间转换成 年月日 类型
    /**<br>gcs:<br>
     * 先把一个long类型的时间yyyy-mm-dd 00:00:00 的格式转换为这样的yyyy-... 的时间格式。之后将转换成功的时间格式的day 提取出来
     * @param longtime 要进行时间转换的long类型的时间
     * @return  将Day返回去的格式是yyyy-MM-dd的格式
     * */
    public String st2Day(Long longtime){
        TimeUtil t = new TimeUtil();
        String day = t.toDay(longtime).substring(0, 10);
        return day;
    }
    //将 st 时间转换成 hour类型
    //将 st 时间转换成 hour类型
    public String st2hour(Long longtime){
        TimeUtil t = new TimeUtil();
        String hours = t.toDay(longtime).substring(11, 13);
        return hours;
    }

    /**<br>gcs:<br>
     *将一个long类型的时间longtime先转换为 "yyyy-MM-dd HH:mm:ss" 的时间格式
     * 之后将格式转换成功的"yyyy-MM-dd HH:mm:ss"的时间
     * @param longtime 要进行时间转换的long类型的时间
     * @return 返回将longtime转换成功后的格式为"yyyy-MM-dd HH:mm:ss"的时间
     * */
    public String toDay(Long longtime){
        Date date=new Date();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        String d = sdf.format(longtime);
        try {
           date=sdf.parse(d);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return d;
    }


    //
    /**<br>gcs:<br>
     * 判断传入的参数是不是日期格式(传参数)   默认获得今天的时间
     * @param args 从-d的参数判断传进来的参数中是否包含-d参数。如果包含-d的参数就把后面的时间提取出来
     * @return 返回"20180625" 这个的时间的格式的String类型的时间
     * */
    public String processArgs(String[] args){

        String date="";
        for (int i=0 ;i<args.length;i++){
            if ("-d".equals(args[i])){ //gcs:如果传进来的args的参数中，有-d这个参数，此时就要把-d后面的时间参数提取出来
                if (i+1 < args.length){
                    date=args[i+1];
                }
            }
        }
        //gcs:判断date是否为Blank，以及查看时间的格式
        if (StringUtils.isNotBlank(date) && chenkDate(date)){
            String[] stDate = date.split("-"); //gcs:将时间按照"-"进行分隔开
            return stDate[0]+stDate[1]+stDate[2]; //gcs:然后将类似于"2018-06-25"的时间，变成"20180625"的时间格式
        }else {
            String tTime = tTime(); //gcs:如果发现date是不符合时间的格式的，此时就会把当前的时间tTime返回来
            return tTime;
        }
    }

    //判断传入的参数是不是日期格式(传参数)  默认昨天
    /**<br>gcs:<br>
     * 将jar包中的args当中的-d 2018-06-25 后面的时间参数2018-06-25 提取出来
     * 如果程序当中没有-d 这个参数，就默认返回"昨天的日期"
     * */
    public String processArgs2(String[] args){
        String date="";
        for (int i=0 ;i<args.length;i++){
            if ("-d".equals(args[i])){
                if (i+1 < args.length){
                    date=args[i+1];
                }
            }
        }
        if (StringUtils.isNotBlank(date) && chenkDate(date)){
            String[] stDate = date.split("-");
            return stDate[0]+stDate[1]+stDate[2];
        }else {
            String tTime = yTime(); //gcs:获得昨天的日期
            return tTime;
        }
    }

    //
    /**<br>gcs:<br>
     * 判断 日期格式是否符合2018-03-23的时间格式
     * @param date 要用来判断格式的String类型的时间
     * @return 如果date符合时间的格式，就会返回来true，否则会返回来false
     * */
    public Boolean chenkDate(String date){
        String regex = "[2][0][0-9]{2}-[0-9]{2}-[0-9]{2}";
        Pattern compile = Pattern.compile(regex);
        //匹配数据
        Matcher matcher = compile.matcher(date);
        return matcher.matches();
    }
    //获得昨天的日期
    /**<br>gcs:<br>
     * 获得昨天的日期
     * */
    public String yTime(){
        Calendar calendar = Calendar.getInstance();
        calendar.add(Calendar.DATE,-1);
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd");
        String s = sdf.format(calendar.getTime());
        return s;
    }

    //获得今天的日期
    public String tTime(){
        Calendar calendar = Calendar.getInstance();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd");
        String s = sdf.format(calendar.getTime());
        return s;
    }
    //获得今天的日期
    public String nowTime(){
        Calendar calendar = Calendar.getInstance();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        String s = sdf.format(calendar.getTime());
        return s;
    }
    //获得今天的 当前小时的前一个小时
    public String getHour(){
        Date now = new Date();
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        String hours= dateFormat.format(now.getTime()-3600000);
        return hours.substring(11, 13);
    }

    /**<br>gcs:<br>
     * 把args当中的参数当中的-h 后面的小时的参数提取出来。如果发现-h后面没有参数，此时就会取得当前的小时的时间
     * @param args 从当期那的参数中把-h后面的小时的参数提取出来
     * @return 将String类型的小时参数返回回来
     * */
    public String processArgsHour(String[] args){
        String hour="";
        for (int i=0 ;i<args.length;i++){
            if ("-h".equals(args[i])){
                if (i+1 < args.length){
                    hour=args[i+1];
                }
            }
        }
        //gcs:如果hour的参数为空
        if (StringUtils.isNotBlank(hour) && chenkHour(hour)){
            return hour; //gcs:如果hour符合格式，此时就会返回String类型的时间
        }else {
            String Mhour = getHour();
            return Mhour; //gcs:否则返回当前的时间
        }
    }

    /**<br>gcs:<br>
     *使用正则表达式，判断小时 05 是否是两位的数字。
     * @param hour 用来判断是否符合格式的hour
     * @return 如果给的hour符合正则表达式，此时就会返回true，否则就会返回false
     * */
    public Boolean chenkHour(String hour){
        String regex = "[0-9]{2}";
        Pattern compile = Pattern.compile(regex);
        //匹配数据
        Matcher matcher = compile.matcher(hour);
        return matcher.matches();
    }


    public static void main(String[] args) {
        TimeUtil t = new TimeUtil();

        //System.out.println(t.chenkak(a));
    }
}
